dev_scripts/checkpoint_eval.py

```diff
@@ -0,0 +1,97 @@
+
+python checkpoint_eval.py "Dev Benchmark 2 32k - 2 Examples" \
+                          "LTMAgentWrapper - gpt-4o - 16384 - 1024" \
+                          "2024-09-03 23:02:43.262463"
+"""
+
+import time_machine
+import argparse
+from argparse import Namespace
+from datetime import datetime, timezone
+from model_interfaces.ltm_agent_wrapper import LTMAgentWrapper
+from utils.files import make_master_log_path
+from runner.master_log import MasterLog, EventType
+from utils.ui import colour_print
+
+
+def get_args() -> Namespace:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("run_name", type=str)
+    parser.add_argument("agent_name", type=str)
+    parser.add_argument("target_timestamp", type=str,
+                        help="The timestamp of the user message that you want to test.")
+    return parser.parse_args()
+
+
+def main(args: Namespace):
+
+    # Set the working time
+    target_date = datetime.fromisoformat(args.target_timestamp)
+    traveller = time_machine.travel(target_date.astimezone(timezone.utc))
+    traveller.start()
+
+    # Load master log and fetch target user message
+    master_log_path = make_master_log_path(args.run_name, args.agent_name)
+    master_log = MasterLog(master_log_path)
+    master_log.load()
+    target_message = None
+    for evt in master_log.log:
+        if evt.timestamp < target_date:
+            continue
+        if evt.type not in {EventType.SEND_MESSAGE, EventType.SEND_FILL}:
+            continue
+        target_message = evt.data["message"]
+        break
+    assert target_message is not None, f"Couldn't find message with timestamp {repr(target_date)}"
+
+    # `model` has "/" symbols removed, but that suffices for loading.
+    model, prompt_size, message_size = args.agent_name.split(" - ")[-3:]
+    agent = LTMAgentWrapper(args.run_name, model=model, max_prompt_size=prompt_size, max_message_size=message_size)
+    agent.load()
+
+    # Patch the retrieval, so that it cannot deliver future memories.
+    def patched_retrieve(*args, **kwargs):
+        memories = patched_retrieve.retrieve(*args, **kwargs)
+        limit_timestamp = target_date.timestamp()
+        return [m for m in memories if m.timestamp < limit_timestamp]
+
+    patched_retrieve.retrieve = agent.agent.hybrid_memory.semantic_memory.retrieve
+    agent.agent.hybrid_memory.semantic_memory.retrieve = patched_retrieve
+
+    # Same for recent messages
+    def patched_recent_messages(*args, **kwargs):
+        messages = patched_recent_messages.get_recent_messages(*args, **kwargs)
+        limit_timestamp = target_date.timestamp()
+        return [m for m in messages if m.timestamp < limit_timestamp]
+
+    patched_recent_messages.get_recent_messages = agent.agent.hybrid_memory.get_recent_messages
+    agent.agent.hybrid_memory.get_recent_messages = patched_recent_messages
```

Comment:
Nice solution!

Reply:
😁 I'm afraid we'll have to fix it at some point, since `retrieve` retrieves a limited amount of memories, and if these happen to be after the filtering date, the result might be quite off. It should work for the moment, though 🤞🏾